steward/src/cork/proposals.rs

```diff
@@ -0,0 +1,320 @@
+use std::time::Duration;
+
+use abscissa_core::{
+    tracing::{debug, error, info, log::warn},
+    Application,
+};
+use gravity_bridge::gravity_proto::{
+    cosmos_sdk_proto::cosmos::base::tendermint::v1beta1::GetLatestBlockRequest,
+    gravity::DelegateKeysByOrchestratorRequest,
+};
+use somm_proto::{
+    cork::ScheduledCorkProposal, cosmos_sdk_proto::cosmos::gov::v1beta1::QueryProposalRequest,
+};
+use steward_proto::steward::{governance_call::Call, GovernanceCall};
+use tokio::task::JoinHandle;
+use tonic::{transport::Channel, Code};
+
+use crate::{
+    cellars::{aave_v2_stablecoin, cellar_v1},
+    config::DELEGATE_ADDRESS,
+    cork::schedule_cork,
+    error::{Error, ErrorKind},
+    prelude::APP,
+};
+
+use super::{client::CorkQueryClient, id_hash};
+
+type GovQueryClient = gravity_bridge::gravity_proto::cosmos_sdk_proto::cosmos::gov::v1beta1::query_client::QueryClient<Channel>;
+type GravityQueryClient =
+    gravity_bridge::gravity_proto::gravity::query_client::QueryClient<Channel>;
+type TendermintQueryClient = gravity_bridge::gravity_proto::cosmos_sdk_proto::cosmos::base::tendermint::v1beta1::service_client::ServiceClient<Channel>;
+
+pub async fn start_scheduled_cork_proposal_polling_thread() -> JoinHandle<()> {
+    debug!("starting cork proposal polling thread");
+    let config = APP.config();
+    let mut state = ProposalThreadState::default();
+    let query_period = Duration::from_secs(config.cork.proposal_poll_period);
+
+    // get validator address that corresponds to delegate for cork scheduling confirmation
+    let mut client = GravityQueryClient::connect(config.cosmos.grpc.clone())
+        .await
+        .expect("failed to connect gravity query client at startup");
+    let request = DelegateKeysByOrchestratorRequest {
+        orchestrator_address: DELEGATE_ADDRESS.to_string(),
+    };
+    state.validator_address = client
+        .delegate_keys_by_orchestrator(request)
+        .await
+        .expect("failed to get validator address at startup.")
+        .into_inner()
+        .validator_address;
+
+    tokio::spawn(async move {
+        let mut fail_count = 0;
+        loop {
+            tokio::time::sleep(query_period).await;
+
+            debug!("polling for new approved scheduled cork proposals");
+            state
+                .update_last_observed_height(config.cosmos.grpc.clone())
+                .await;
+            if let Err(err) = poll_approved_cork_proposals(&mut state).await {
+                fail_count += 1;
+                error!(
+                    "proposal {} failed to process {} time(s). latest error: {}",
+                    state.last_processed_proposal_id + 1,
+                    fail_count,
+                    err
+                );
+            } else {
+                fail_count = 0;
+            }
+        }
+    })
+}
+
+#[derive(Debug, Default)]
+struct ProposalThreadState {
+    last_observed_height: u64,
+    last_processed_proposal_id: u64,
```

Comment:
I guess the default value of a u64 is zero then, so we don't need to be explicit for the base case.